gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemBase.java

```diff
@@ -1503,25 +1503,28 @@ private Credential getCredential(Configuration config)
    */
   private static Optional<Credential> getImpersonatedCredential(
       Configuration config, Credential credential) throws IOException {
-    String serviceAccountToImpersonate =
-        IMPERSONATION_SERVICE_ACCOUNT_SUFFIX
-            .withPrefixes(CONFIG_KEY_PREFIXES)
-            .get(config, config::get);
-
     GoogleCloudStorageOptions options =
         GoogleHadoopFileSystemConfiguration.getGcsFsOptionsBuilder(config)
             .build()
             .getCloudStorageOptions();
 
-    serviceAccountToImpersonate =
+    String serviceAccountFromUserOrGroup =
         getMatchedServiceAccountToImpersonateFromGroup(
                 options.getUserImpersonationServiceAccounts(),
                 ImmutableList.of(UserGroupInformation.getCurrentUser().getShortUserName()))
             .orElse(
                 getMatchedServiceAccountToImpersonateFromGroup(
                         options.getGroupImpersonationServiceAccounts(),
                         ImmutableList.copyOf(UserGroupInformation.getCurrentUser().getGroupNames()))
-                    .orElse(serviceAccountToImpersonate));
+                    .orElse(null));
+
+    String serviceAccountToImpersonate =
+        IMPERSONATION_SERVICE_ACCOUNT_SUFFIX
+            .withPrefixes(CONFIG_KEY_PREFIXES)
+            .get(config, config::get);
```

Comment:
This is actually not matter of readability, it will change the logic to make this least priority. Current logic is to let the SA specified by impersonation.service.account supersede the ones specified in mappings.